### PR TITLE
SAT solver output parser

### PR DESCRIPTION
**Pull request by atimaly**
- **SAT solver output parser**
- **SAT solver output parser**

# Description of the Contribution

# PR Checklist

- [x] I read and agree to [`CONTRIBUTING.md`](https://github.com/chrjabs/rustsat/blob/main/CONTRIBUTING.md)
- [x] I have formatted my code with `rustfmt` / `cargo fmt --all`
- [ ] Commits are named following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [ ] I have added documentation for new features
- [x] The test suite still passes on this PR
- [ ] I have added tests for new features / tests that would have caught the bug this PR fixes (please explain if not)

### CODE REVIEW

File: rustsat/src/types.rs [hunk 1]
```diff
@@ -697,6 +701,39 @@ impl Assignment {
...
+            .context("failed to open reader")?;
+        match instances::fio::parse_sat_solver_output(std::io::BufReader::new(reader)) {
+            Ok(instances::fio::SolverOutput::Sat(assignment)) => Ok(assignment),
+            _ => Err(anyhow::anyhow!(
+                instances::fio::SatSolverOutputError::Nonsolution
+            )), // return some error
+        }
+    }
+
+    pub fn from_vline(&mut self, line: &str) {
```
Comment (suomela) [inline rustsat/src/types.rs:716]: Should return `Result<()>`?

File: rustsat/src/types.rs [hunk 2]
```diff
@@ -697,6 +701,39 @@ impl Assignment {
...
+                number_v = number.parse::<i32>().unwrap();
+            }
+            //End of the value lines
+            if number_v == 0 {
+                continue;
+            }
+
+            let literal = match Lit::from_ipasir(number_v) {
+                Ok(lit) => lit,
+                Err(error) => panic!("Error creating the literal {:?}", error),
```
Comment (suomela) [inline rustsat/src/types.rs:731]: Should not panic but return the error?

File: rustsat/src/instances/fio.rs [hunk 3]
```diff
@@ -66,3 +72,56 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
...
+}
+
+pub fn parse_sat_solver_output<R: BufRead>(reader: R) -> anyhow::Result<SolverOutput> {
+    // Parsing code goes here
+    // when encoutering vline, call `Assignment::from_vline`
+
+    let mut sat_solver_output = SolverOutput::Sat(Default::default());
+
+    for line in reader.lines() {
+        let line_ls = line.as_ref().expect("Couldn't read SAT solution line");
```
Comment (suomela) [inline rustsat/src/instances/fio.rs:103]: Should not panic but should return error?

File: rustsat/src/types.rs [hunk 4]
```diff
@@ -697,6 +701,41 @@ impl Assignment {
...
+
+            let literal = match Lit::from_ipasir(number_v) {
+                Ok(lit) => lit,
+                Err(error) => return Err(anyhow::anyhow!(error)),
+            };
+
+            self.assign_lit(literal);
+        }
+
+        return Ok(());
```
Comment (chrjabs) [inline rustsat/src/types.rs:737]: I'd prefer this to be a constructor following Rust's `from_` naming scheme
```suggestion
    pub fn from_vline(line: &str) -> anyhow::Result<Self> {
        // skip `v ` tag
        let line = line[2:];
        let assignment = Assignment::default();
        for number in line.split(' ') {
            let number = number.parse::<i32>()?;
            //End of the value lines
            if number == 0 {
                return assignment;
            }
            let literal = Lit::from_ipasir(number)?;
            assignment.assign_lit(literal);
        }

        return Ok(());
```

**Review by chrjabs — CHANGES_REQUESTED** [review]
Here is some overall comments. Feel free to ask if you have any questions regarding what I mean. Also, I'm writing these suggestions without checking whether they compile, so take them more as ideas than as exact changes to make.

File: rustsat/src/instances/fio.rs [hunk 5]
```diff
@@ -66,3 +72,60 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
...
+        }
+
+        //Value line
+        if line_ls.starts_with('v') {
+            let mut current_assignment = match sat_solver_output {
+                SolverOutput::Sat(ref mut assign) => assign,
+                _ => return Err(anyhow::anyhow!(SatSolverOutputError::Nonsolution)),
+            };
+
+            current_assignment.from_vline(&line_ls)?;
```
[the remark this answers is not in the document]
Comment (atimaly) [reply rustsat/src/instances/fio.rs:126]: The problem I had with creating a constructor is that there can be multiple v lines and so I did not see how to solve this with the same function.

[the remark this answers is not in the document]
Comment (chrjabs) [reply rustsat/src/instances/fio.rs:126]: Right, I forgot about the fact that the assignment might be split over multiple lines. I'll have to think about this a bit, will probably get back with an idea next week.

[the remark this answers is not in the document]
Comment (chrjabs) [reply rustsat/src/instances/fio.rs:126]: After thinking about this some more, I still think that for usability and intuitiveness, `from_vline` should be a constructor. I'd therefore suggest to introduce a second method `extend_from_vline` that does what your current implementation does. The constructor can then simply initialize a default empty assignment and call `extend_from_vline`, but it's still available for useability.

**Comment by chrjabs** [comment]
Another question to keep generally in mind for this is what to treat as an actual error and what to accept even though it might not technically be according to the specification. I try to implement my parsers to still accept things that "make sense". For example if an assignment here is not terminated by a zero, I think that's worth still accepting. On the other hand, if an assignment has conflicting literals, this parser shouldn't silently choose one but actually fail.

File: rustsat/src/instances/fio.rs [hunk 6]
```diff
@@ -66,3 +75,82 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
...
+            return Ok(SolverOutput::Sat(solution));
+        }
+    }
+
+    if is_sat {
+        Err(anyhow::anyhow!(SatSolverOutputError::NoSline))
+    } else {
+        Err(anyhow::anyhow!(SatSolverOutputError::NoVline))
+    }
+}
```
Comment (chrjabs) [inline rustsat/src/instances/fio.rs:156]:
```suggestion
}


#[cfg(test)]
mod tests {
    use std::io;

    use crate::types::{Assignment, TernaryVal};

    use super::{parse_sat_solver_output, SolverOutput};

    #[test]
    fn parse_solver_output_sat() {
        let ground_truth = SolverOutput::Sat(Assignment::from(vec![
            TernaryVal::True,
            TernaryVal::False,
            TernaryVal::DontCare,
            TernaryVal::True,
            TernaryVal::False,
            TernaryVal::True,
        ]));

        let data = "c this is a comment\ns SATISFIABLE\nv 1 -2 4 -5 6 0\n";
        let reader = io::Cursor::new(data);
        let res = parse_sat_solver_output(reader).unwrap();
        assert_eq!(res, ground_truth);

        let data = "c this is a comment\nv 1 -2 4 -5 6 0\ns SATISFIABLE\n";
        let reader = io::Cursor::new(data);
        let res = parse_sat_solver_output(reader).unwrap();
        assert_eq!(res, ground_truth);

        let data = "c this is a comment\ns SATISFIABLE\nv 1 -2 4 \nv -5 6 0\n";
        let reader = io::Cursor::new(data);
        let res = parse_sat_solver_output(reader).unwrap();
        assert_eq!(res, ground_truth);
    }

    #[test]
    fn parse_solver_output_unsat() {
        let data = "c this is a comment\ns UNSATISFIABLE\n";
        let reader = io::Cursor::new(data);
        let res = parse_sat_solver_output(reader).unwrap();
        assert_eq!(res, SolverOutput::Unsat);
    }

    #[test]
    fn parse_solver_output_unknown() {
        let data = "c this is a comment\ns UNKNOWN\n";
        let reader = io::Cursor::new(data);
        let res = parse_sat_solver_output(reader).unwrap();
        assert_eq!(res, SolverOutput::Unknown);
    }
}
```

File: rustsat/src/instances/fio.rs [hunk 7]
```diff
@@ -66,3 +75,82 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
     }
     Ok(Box::new(io::BufWriter::new(raw_writer)))
 }
+
+pub enum SolverOutput {
```
Comment (chrjabs) [inline rustsat/src/instances/fio.rs:79]:
```suggestion
#[derive(Debug, PartialEq, Eq)]
pub enum SolverOutput {
```

**Review by chrjabs — CHANGES_REQUESTED** [review]
Before I give more implementation specific comments, could you please add these basic tests and make sure they pass, so that I don't have to worry about whether the parsing logic is correct when looking at the code? Thanks.

Here are some additional tests that should at the bottom of the `types` module.
```rust

    #[test]
    fn parse_vline() {
        let vline = "v 1 -2 4 -5 6 0";
        let ground_truth = Assignment::from(vec![
            TernaryVal::True,
            TernaryVal::False,
            TernaryVal::DontCare,
            TernaryVal::True,
            TernaryVal::False,
            TernaryVal::True,
        ]);
        assert_eq!(Assignment::from_vline(vline).unwrap(), ground_truth);
        assert_eq!(
            {
                let mut asign = Assignment::default();
                asign.extend_from_vline(vline).unwrap();
                asign
            },
            ground_truth
        );
    }

    #[test]
    #[should_panic]
    fn vline_invalid_lit_from() {
        let vline = "v 1 -2 4 foo -5 bar 6 0";
        Assignment::from_vline(vline).unwrap();
    }

    #[test]
    #[should_panic]
    fn vline_invalid_lit_extend() {
        let vline = "v 1 -2 4 foo -5 bar 6 0";
        let mut assign = Assignment::default();
        assign.extend_from_vline(vline).unwrap();
    }

    #[test]
    #[should_panic]
    fn vline_invalid_tag_from() {
        let vline = "b 1 -2 4 -5 6 0";
        Assignment::from_vline(vline).unwrap();
    }

    #[test]
    #[should_panic]
    fn vline_invalid_tag_extend() {
        let vline = "b 1 -2 4 -5 6 0";
        let mut assign = Assignment::default();
        assign.extend_from_vline(vline).unwrap();
    }
```

File: rustsat/src/instances/fio.rs (hunk 6 above)
Comment (ChrisJefferson) [reply rustsat/src/instances/fio.rs:156]: Just some more suggested tests I've messed up myself over the years:

A completely empty SAT file (which people sometimes generate automatically, so should be handled cleanly): `s SATISFIABLE\nv 0\n`

**Comment by ChrisJefferson** [comment]
Some suggestions from parsing a bunch of SAT output (I'd add tests for all of these):

Some solvers output INDETERMINATE instead of UNKNOWN (I'd treat them exactly the same, in both cases the solver probably ran out of time, or some other limit). For an example discussion, https://github.com/arminbiere/cadical/issues/18

When solvers get very confused they won't output an `s` line at all. If there is no `s` line, even if there are `v` lines, output the solver errored because you can't be sure what happened.

**Comment by atimaly** [comment]
> Before I give more implementation specific comments, could you please add these basic tests and make sure they pass, so that I don't have to worry about whether the parsing logic is correct when looking at the code? Thanks.
> 
> Here are some additional tests that should at the bottom of the `types` module.
> 
> ```rust
>     #[test]
>     fn parse_vline() {
>         let vline = "v 1 -2 4 -5 6 0";
>         let ground_truth = Assignment::from(vec![
>             TernaryVal::True,
>             TernaryVal::False,
>             TernaryVal::DontCare,
>             TernaryVal::True,
>             TernaryVal::False,
>             TernaryVal::True,
>         ]);
>         assert_eq!(Assignment::from_vline(vline).unwrap(), ground_truth);
>         assert_eq!(
>             {
>                 let mut asign = Assignment::default();
>                 asign.extend_from_vline(vline).unwrap();
>                 asign
>             },
>             ground_truth
>         );
>     }
> 
>     #[test]
>     #[should_panic]
>     fn vline_invalid_lit_from() {
>         let vline = "v 1 -2 4 foo -5 bar 6 0";
>         Assignment::from_vline(vline).unwrap();
>     }
> 
>     #[test]
>     #[should_panic]
>     fn vline_invalid_lit_extend() {
>         let vline = "v 1 -2 4 foo -5 bar 6 0";
>         let mut assign = Assignment::default();
>         assign.extend_from_vline(vline).unwrap();
>     }
> 
>     #[test]
>     #[should_panic]
>     fn vline_invalid_tag_from() {
>         let vline = "b 1 -2 4 -5 6 0";
>         Assignment::from_vline(vline).unwrap();
>     }
> 
>     #[test]
>     #[should_panic]
>     fn vline_invalid_tag_extend() {
>         let vline = "b 1 -2 4 -5 6 0";
>         let mut assign = Assignment::default();
>         assign.extend_from_vline(vline).unwrap();
>     }
> ```

Currently I have added both yours (@chrjabs ), @ChrisJefferson's and mine tests. All of them are passed.

**Comment by chrjabs** [comment]
Looks like the tests I suggested for `rustsat/src/types.rs` are still missing?

**Comment by atimaly** [comment]
> Looks like the tests I suggested for `rustsat/src/types.rs` are still missing?

Sorry forgot about that part.

**Comment by ChrisJefferson** [comment]
Couple of comments (and note I'm just some person, so obviously don't do what I say if it disagrees with @chrjabs ! )

I don't know if a solver would produce it, but you don't handle:

"c this is a comment\nv 1 -2 4 \ns SATISFIABLE\nv -5 6 0\n";

Because you return early if you have some solution. I'd personally just never return from `line.contains("SATISFIABLE")`, just remember `is_sat`, then gather up the solution at the end.

Also, sorry for poorly explaining, I don't think you should error if there are no returned values, having an empty SAT instance is perfectly fine!, just return SAT, and no variables. I just wanted to check that case didn't break.

Also, I think lines 146-163 are a bit messy. I think the cases are:

if !is_sat: In this case something bad has happened, error (as on 146-148 currently, this is right).

After this, we should have a vline (as we know we are satisfiable), so you can just check if you have a solution (if so return it), or if you don't, return NoVline?

**Comment by atimaly** [comment]
> Looks like the tests I suggested for `rustsat/src/types.rs` are still missing?

Added and passed.

**Comment by chrjabs** [comment]
I'll go through this in detail again in the next days. If you (@atimaly) have time before that, feel free to clean up the logic a bit and possibly write some documentation already. From a quick look, I also agree with the comment from @ChrisJefferson, these are exactly the edge cases I was talking about that should be handled "reasonably". You could also run Minisat/glucose/cadical/kissat on an instance and use their output files in tests.

File: rustsat/src/instances/fio.rs [hunk 8]
```diff
@@ -66,3 +75,167 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
...
+                    return Ok(SolverOutput::Unknown)
+                }
+                line if line.contains("SATISFIABLE") => {
+                    /*if let Some(solution) = solution {
+                        return Ok(SolverOutput::Sat(solution));
+                    }*/
+                    is_sat = true;
+                }
+                _ => return Err(anyhow::anyhow!(SatSolverOutputError::InvalidSLine)),
+            }
```
Comment (chrjabs) [inline rustsat/src/instances/fio.rs:121]:
```suggestion
            // skip s and trailing whitespace
            let line = &line[1..].trim_start();
            match line {
                line if line.starts_with("UNSATISFIABLE") => return Ok(SolverOutput::Unsat),
                line if line.starts_with("UNKNOWN") || line.starts_with("INDETERMINATE") => {
                    return Ok(SolverOutput::Unknown)
                }
                line if line.starts_with("SATISFIABLE") => {
                    is_sat = true;
                }
                _ => anyhow::bail!(SatSolverOutputError::InvalidSLine),
            }
```
I think slightly more thorough checks here are reasonable. Note how your version would accept `s ERROR UNSATISFIABLE` as a valid Unsat marker.

File: rustsat/src/instances/fio.rs [hunk 9]
```diff
@@ -66,3 +75,167 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
...
+                    Some(ref mut assign) => assign,
+                    _ => return Err(anyhow::anyhow!(SatSolverOutputError::Nonsolution)),
+                };
+
+                current_assignment.extend_from_vline(&line)?;
+            } else {
+                solution = Some(Assignment::from_vline(&line)?);
+            }
+
+            have_vline = true;
```
Comment (chrjabs) [inline rustsat/src/instances/fio.rs:138]:
```suggestion
            match &mut solution {
                Some(assign) => assign.extend_from_vline(&line)?,
                _ => solution = Some(Assignment::from_vline(&line)?,
            }
```

File: rustsat/src/instances/fio.rs [hunk 10]
```diff
@@ -66,3 +75,167 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
...
+        }
+    }
+
+    //There is no solution line so we can not trust the output
+    if !is_sat {
+        return Err(anyhow::anyhow!(SatSolverOutputError::NoSline));
```
Comment (chrjabs) [inline rustsat/src/instances/fio.rs:144]:
```suggestion
        anyhow::bail!(SatSolverOutputError::NoSline);
```
I just learned about this macro. This makes the expression a bit simpler.

File: rustsat/src/types.rs [hunk 11]
```diff
@@ -697,6 +704,59 @@ impl Assignment {
...
+            assignment.assign_lit(literal);
+        }
+
+        Ok(assignment)
+    }
+
+    /// Parses and saves literals from value line.
+    pub fn extend_from_vline(&mut self, line: &str) -> anyhow::Result<()> {
+        if !line.starts_with('v') {
+            panic!("Value line does not start with v.");
```
Comment (chrjabs) [inline rustsat/src/types.rs:742]: Sorry in case the `#[should_panic]` on the tests was confusing. As a rule of thumb, never use `panic` inside the library, but always return errors. Once your error types are fixed, we can update the tests to expect the exact error rather than simply panic.
For here, please introduce a `InvalidVLine` enum with at least the variants `InvalidTag(char)` and `ConflictingAssignment(Var)` and use them appropriately.

File: rustsat/src/types.rs [hunk 12]
```diff
@@ -697,6 +704,59 @@ impl Assignment {
...
+            //End of the value lines
+            if number_v == 0 {
+                continue;
+            }
+
+            let literal = Lit::from_ipasir(number_v)?;
+            assignment.assign_lit(literal);
+        }
+
+        Ok(assignment)
```
Comment (chrjabs) [inline rustsat/src/types.rs:736]:
```suggestion
        let mut assignment = Assignment::default();
        assignment.extend_from_vline(line)?;
        Ok(assignment)
```
This way the parsing logic doesn't have to be duplicated.

File: rustsat/src/types.rs [hunk 13]
```diff
@@ -697,6 +704,59 @@ impl Assignment {
...
+    }
+
+    /// Parses and saves literals from value line.
+    pub fn extend_from_vline(&mut self, line: &str) -> anyhow::Result<()> {
+        if !line.starts_with('v') {
```
Comment (chrjabs) [inline rustsat/src/types.rs:741]:
```suggestion
        if !line.starts_with("v ") {
```
Otherwise `vfoo` would be counted as a valid starting tag.

File: rustsat/src/types.rs [hunk 14]
```diff
@@ -697,6 +704,59 @@ impl Assignment {
...
+            panic!("Value line does not start with v.");
+        }
+
+        let line = &line[2..];
+        for number in line.trim().split(' ').filter(|x| !x.is_empty()) {
+            let number_v = number.parse::<i32>().unwrap();
+
+            //End of the value lines
+            if number_v == 0 {
```
Comment (chrjabs) [inline rustsat/src/types.rs:750]:
```suggestion
            if number == 0 {
```

File: rustsat/src/types.rs [hunk 15]
```diff
@@ -697,6 +704,59 @@ impl Assignment {
...
+                continue;
+            }
+
+            let literal = Lit::from_ipasir(number_v)?;
+            self.assign_lit(literal);
```
Comment (chrjabs) [inline rustsat/src/types.rs:755]:
```suggestion
            let literal = Lit::from_ipasir(number)?;
            let val = self.lit_value(literal);
            if val == TernaryVal::True && literal.is_neg() || val == TernaryVal::False && literal.is_pos() {
                // Catch conflicting assignments
                anyhow::bail!(InvalidVLine::ConflictingAssignment(literal.var());
            }
            self.assign_lit(literal);
```

File: rustsat/src/types.rs [hunk 16]
```diff
@@ -697,6 +704,59 @@ impl Assignment {
...
+        }
+
+        return Ok(());
```
Comment (chrjabs) [inline rustsat/src/types.rs:758]:
```suggestion
        Ok(())
```

File: rustsat/src/instances/fio.rs [hunk 17]
```diff
@@ -66,3 +75,167 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
...
+        let res = parse_sat_solver_output(reader).unwrap();
+        assert_eq!(res, SolverOutput::Unknown);
+    }
+
+    #[test]
+    fn parse_solver_output_noslinewithvline() {
+        let data = "c this is a comment\nv 1 -2 4 -5 6 0\n";
+        let reader = io::Cursor::new(data);
+        let res = parse_sat_solver_output(reader);
+        assert!(res.is_err());
```
Comment (chrjabs) [inline rustsat/src/instances/fio.rs:223]:
```suggestion
        match res.downcast::<SolverOutputError>() {
            Some(err) => match err {
                SolverOutputError::NoSLine => (),
                _ => panic!()
            }
            Err(_) => panic!(),
        }
```
This (or something along those lines, I'm writing this free hand) is how you can check for the exact error type in the anyhow error. It would be great if you can do this for the other error cases as well.

File: rustsat/src/types.rs [hunk 18]
```diff
@@ -981,4 +1041,56 @@ mod tests {
...
+        Assignment::from_vline(vline).unwrap();
+    }
+
+    #[test]
+    #[should_panic]
+    fn vline_invalid_tag_extend() {
+        let vline = "b 1 -2 4 -5 6 0";
+        let mut assign = Assignment::default();
+        assign.extend_from_vline(vline).unwrap();
+    }
```
Comment (chrjabs) [inline rustsat/src/types.rs:1095]: Once you have the `InvalidVLine` error type implemented, feel free to change these to check for the exact error with `downcast` as I mentioned above.

File: rustsat/src/instances/fio.rs [hunk 19]
```diff
@@ -66,3 +75,167 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
...
+                line if line.contains("SATISFIABLE") => {
+                    /*if let Some(solution) = solution {
+                        return Ok(SolverOutput::Sat(solution));
+                    }*/
+                    is_sat = true;
+                }
+                _ => return Err(anyhow::anyhow!(SatSolverOutputError::InvalidSLine)),
+            }
+        }
+
```
Comment (chrjabs) [inline rustsat/src/instances/fio.rs:123]:
```suggestion

    if line.starts_with("c s UNKNOWN") {
        // special case for CaDiCaL
        return Ok(SolverOutput::Unknown);
    }

```
@ChrisJefferson, based on the CaDiCaL discussion you linked, do you think catching this special case makes sense?

**Review by chrjabs — CHANGES_REQUESTED** [review]
I think we're getting there, this is already significantly more solid code than the first version. I hope you can see why I am relatively picky about catching edge cases we can think of here, if this code is supposed to go into this library I would like the problems I can already see to be worked out.
Thanks for your work on this :+1:

**Comment by chrjabs** [comment]
Also, don't worry about the failing macos test, that one has some weird flakeyness that I haven't figured out how to avoid yet.

**Comment by atimaly** [comment]
> I think we're getting there, this is already significantly more solid code than the first version. I hope you can see why I am relatively picky about catching edge cases we can think of here, if this code is supposed to go into this library I would like the problems I can already see to be worked out. Thanks for your work on this 👍

I am happy that we are getting there. Also I am really thankful for the comments and suggestions. I fully understand why you would be picky about the edge cases. The edge case that I want to handle most is the case where there is the same literal but with different sign in the solution.

File: rustsat/src/types.rs [hunk 20]
```diff
@@ -697,6 +704,59 @@ impl Assignment {
...
+
+    /// Parses and saves literals from value line.
+    pub fn extend_from_vline(&mut self, line: &str) -> anyhow::Result<()> {
+        if !line.starts_with('v') {
+            panic!("Value line does not start with v.");
+        }
+
+        let line = &line[2..];
+        for number in line.trim().split(' ').filter(|x| !x.is_empty()) {
+            let number_v = number.parse::<i32>().unwrap();
```
[the remark this answers is not in the document]
Comment (atimaly) [reply rustsat/src/types.rs:747]: because of the tests. I have seen should_panic.

File: rustsat/src/types.rs (hunk 15 above)
Comment (atimaly) [reply rustsat/src/types.rs:755]: Oh, I see you already catch this edge case :).

File: rustsat/src/types.rs [hunk 21]
```diff
@@ -697,6 +704,56 @@ impl Assignment {
...
+    pub fn from_solver_output_path<P: AsRef<Path>>(path: P) -> anyhow::Result<SolverOutput> {
+        let reader = std::io::BufReader::new(
+            instances::fio::open_compressed_uncompressed_read(path)
+                .context("failed to open reader")?,
+        );
+        instances::fio::parse_sat_solver_output(reader)
+    }
+
+    /// Given a value line of a SAT solver's solution create an Assignment with
+    /// this line's literal values already included
```
Comment (chrjabs) [inline rustsat/src/types.rs:718]:
```suggestion
    /// Creates an assignment from a SAT solver value line
```

File: rustsat/src/types.rs [hunk 22]
```diff
@@ -697,6 +704,56 @@ impl Assignment {
...
+        if !line.starts_with("v ") {
+            if line.is_empty() {
+                anyhow::bail!(InvalidVLine::Emptyline);
+            }
+
+            anyhow::bail!(InvalidVLine::InvalidTag(line.chars().next().unwrap()));
+        }
+
+        let line = &line[1..].trim_start();
+        for number in line.trim().split(' ').filter(|x| !x.is_empty()) {
```
Comment (chrjabs) [inline rustsat/src/types.rs:736]:
```suggestion
        let line = &line[1..];
        for number in line.split_whitespace() {
```
`split_whitespace` automatically removes empty elements and basically performs the trim operation.

File: rustsat/src/types.rs [hunk 23]
```diff
@@ -697,6 +704,56 @@ impl Assignment {
...
+    /// this line's literal values already included
+    pub fn from_vline(line: &str) -> anyhow::Result<Self> {
+        let mut assignment = Assignment::default();
+        assignment.extend_from_vline(line)?;
+        Ok(assignment)
+    }
+
+    /// Parses and saves literals from a value line.
+    pub fn extend_from_vline(&mut self, line: &str) -> anyhow::Result<()> {
+        if !line.starts_with("v ") {
```
Comment (chrjabs) [inline rustsat/src/types.rs:727]:
```suggestion
        for line in line.lines() {
            if !line.starts_with("v ") {
```
I just realized that a user could call this function with a huge string slice that already contains multiple v lines. Even though I would not recommend this, I think it might make sense to still support this usecase. (Of course close the for loop at the end and indent appropriately.)

File: rustsat/src/types.rs [hunk 24]
```diff
@@ -981,4 +1040,93 @@ mod tests {
...
+        let vline = "";
+        let res = Assignment::from_vline(vline);
+        match res.unwrap_err().downcast::<InvalidVLine>() {
+            Ok(err) => match err {
+                InvalidVLine::Emptyline => assert!(true),
+                _ => panic!(),
+            },
+            Err(_) => panic!(),
+        }
+    }
```
Comment (chrjabs) [inline rustsat/src/types.rs:1131]:
```suggestion
    }
    
    #[test]
    fn multi_vline() {
        let vline = "v 1 2 3\nv 4 5 6 0";
                let ground_truth = Assignment::from(vec![
            TernaryVal::True,
            TernaryVal::True,
            TernaryVal::True,
            TernaryVal::True,
            TernaryVal::True,
            TernaryVal::True,
        ]);
        let res = Assignment::from_vline(vline)?;
        assert_eq!(res, ground_truth);
    }
```

File: rustsat/src/instances/fio.rs [hunk 25]
```diff
@@ -66,3 +75,179 @@ pub(crate) fn open_compressed_uncompressed_write<P: AsRef<Path>>(
...
+/// The possible errors that can happen to a value line in the SAT solver's output.
+#[derive(Error, Debug)]
+pub enum InvalidVLine {
+    #[error("The value line does not start with 'v ' ")]
+    InvalidTag(char),
+    #[error("The output of the SAT solver assigned the same variable different values.")]
+    ConflictingAssignment(types::Var),
+    #[error("Empty value line.")]
+    Emptyline,
```
Comment (chrjabs) [inline rustsat/src/instances/fio.rs:108]:
```suggestion
    #[error("The value line does not start with 'v ' but with '{}'")]
    InvalidTag(char),
    #[error("The output of the SAT solver assigned different values to variable {}")]
    ConflictingAssignment(types::Var),
    #[error("Empty value line")]
    EmptyLine,
```

**Comment by chrjabs** [comment]
Ah, btw, I'll most likely squash this PR in the end into a single commit anyway, so don't worry about conventional commits right now.

**Comment by atimaly** [comment]
> Can you please still include these changes. After that, the last thing I see missing right now are some tests with actual solver output files. So maybe you could run some solvers, e.g., CaDiCaL / MiniSAT on the [this instance](https://github.com/chrjabs/rustsat/blob/main/data/AProVE11-12.cnf) and [this instance](https://github.com/chrjabs/rustsat/blob/main/data/smtlib-qfbv-aigs-ext_con_032_008_0256-tseitin.cnf), put the solver output files in the `data/` directory and include tests for `fio::parse_sat_solver_output` on those files. That would be great.

I have been trying to make this but for some reason I get back the following message "the trait bound `rustsat_minisat::core::Minisat: Solve` is not satisfied" for both of the codes below.
`use crate::{
            instances::{BasicVarManager, SatInstance},
            solvers,
            solvers::SolverResult,
        };
            let inst: SatInstance<BasicVarManager> =
                SatInstance::from_dimacs_path("./data/AProVE11-12.cnf.gz").unwrap();
            let mut solver = rustsat_minisat::core::Minisat::default();
            solvers::Solve::add_cnf(&mut solver, inst.as_cnf().0);
            //solver.add_cnf(inst.into_cnf().0).unwrap();
            let res = solver.solve().unwrap();
            assert_eq!(res, SolverResult::Sat);`

`use crate::{
            instances::{SatInstance},
            lit,
            solvers::{Solve, SolveIncremental, SolverResult::{Sat, Unsat}},
        };
        
        let mut instance = SatInstance::<instances::BasicVarManager>::from_dimacs_path("../../data/AProVE11-12.cnf").unwrap();
        let mut solver = rustsat_minisat::core::Minisat::default();
        super::super::super::solvers::Solve::add_cnf(&mut solver, instance.as_cnf().0);
        //super::solvers::Solve::add_cnf(instance.as_cnf().0).unwrap();
        let res = solver.solve().unwrap();`

**Comment by chrjabs** [comment]
> I have been trying to make this but for some reason I get back the following message "the trait bound `rustsat_minisat::core::Minisat: Solve` is not satisfied" for both of the codes below.

What are you trying to use this code for? In the tests?
What I meant you should do is use a separately compiled binary of Minisat (e.g., from [this repo](https://github.com/chrjabs/minisat)) and CaDiCal (from [this repo](https://github.com/arminbiere/cadical)) and run those binaries as a CLI tool on the instances in `data/` and collect their output. This is entirely separate from Rust and simulates the intended usecase of running an external solver that is not included in RustSAT. You can then use the collected solver output file in tests.

**Comment by chrjabs** [comment]
I think this looks pretty good now. I'll merge this next week, but probably manually instead of on GitHub through this PR. Are you (@atimaly) fine with me removing the authorship of @atimalyfreiburg, which I assume is just another email address of yours?

**Comment by atimaly** [comment]
> I think this looks pretty good now. I'll merge this next week, but probably manually instead of on GitHub through this PR. Are you (@atimaly) fine with me removing the authorship of @atimalyfreiburg, which I assume is just another email address of yours?

Yeah, I am fine with removing the @atimalyfreiburg authorship (university email).
